lib/index.js

```diff
@@ -112,15 +125,31 @@ function doDownload(stream, url, options, tryCount) {
     if (res.statusCode !== 200 && res.statusCode !== 206) {
       if (redirectCodes.has(res.statusCode)) {
         // Redirection header.
-        doDownload(stream, res.headers.location, options, tryCount - 1);
+        doDownload(stream, res.headers.location, options, tryCount - 1, reconnectInfo);
         return;
       }
       stream.emit('response', res);
       stream.emit('error', new Error('Status code ' + res.statusCode));
       return;
     }
 
-    res.pipe(stream);
+    reconnectInfo.range.end = reconnectInfo.range.end === -1 ? parseInt(res.headers['content-length']) : (parseInt(res.headers['content-length']) < reconnectInfo.range.end ? parseInt(res.headers['content-length']) : reconnectInfo.range.end);
+    res.on('data', function(d) {
+      reconnectInfo.downloaded += d.length;
+      stream.emit('progress', reconnectInfo.downloaded / (reconnectInfo.range.end - reconnectInfo.range.start), reconnectInfo.range.start, reconnectInfo.range.end, reconnectInfo.downloaded, d.length);
+    });
+
+    res.pipe(stream, {end : false});  //not sure wether this pipes errors...
+    res.on('end', function() {
+      stream.unpipe(res);
+      if (reconnectInfo.downloaded + reconnectInfo.range.start < reconnectInfo.range.end) {
```

Comment:
I had thought that the reason `end` was being listened to was because it was expected to happen when there stream suddenly disconnects mid-download. But if it does disconnect, an `error` event would be emitted, but not an `end` event.